### PR TITLE
Build Lego (ELM Basic Interface)

### PR DESCRIPTION
**Pull request by herbiebradley**
This PR implements ELM stage 1 using Salesforce CodeGen models and prompt engineering.

Feedback welcomed on the interface/code structure. The current state is very untested and not cleaned up, this PR is being submitted now to show the code structure.

### CODE REVIEW

File: walker/walk_creator.py [hunk 1]
```diff
@@ -37,9 +39,38 @@ def serialize_walker(self):
...
+                walker_dict['connections'].append({
+                    "sourceID": str(self.joint_index(m.j0)),
+                    "targetID": str(self.joint_index(m.j1)),
+                    "cppnOutputs": [0, 0, 0, -10.0]
+                })
+            elif m.type == "muscle":
+                walker_dict['connections'].append({
+                    "sourceID": str(self.joint_index(m.j0)),
+                    "targetID": str(self.joint_index(m.j1)),
+                    "cppnOutputs": [0, 0, m.phase, m.amplitude]
```
Comment (herbiebradley) [inline walker/walk_creator.py:70]: @dmahan93 Just making sure this makes sense to construct a dict like the example ones you provided. The code does not use the first two indexes of `cppnOutputs`, right, assuming `useLEO=True`? Also do you know the purpose of the `useLEO` option?

Comment (dmahan93) [reply walker/walk_creator.py:70]: @herbiebradley That's correct yeah. The useLEO option is a hyperneat thing, (pg. 3 http://eplex.cs.ucf.edu/papers/szerlip_ecal13.pdf), so we can safely ignore it.

File: map_elites/environments.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
 import math
 import string
+from abc import ABC, abstractmethod
 
 import numpy as np
 from numpy import array
 
 from map_elites.map_elites import Genotype, Phenotype
```
Comment (honglu2875) [inline map_elites/environments.py:8]: There was a problem of circular import when I tried to run `elm.py` from the previous directory. Was using Python 3.9.
Also, judging from the next two lines, are you running something from this directory instead?

File: elm.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+# Load hydra config from yaml files and command line arguments.
+@hydra.main(config_path=str(PROJECT_PATH), config_name="elm_cfg",
+            version_base="1.2")
+def main(cfg):
+    print('----------------- Config ---------------')
+    print(OmegaConf.to_yaml(cfg))
+    print('-----------------  End -----------------')
+    elm = ELM(cfg)
+    print("Best Sodaracer: ", elm.run)
```
Comment (honglu2875) [inline elm.py:80]: Guess it's a typo here? `..., elm.run())`

File: diff_model.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            encoding = tokenizer([seed], truncation=True, padding=True,
+                                max_length=2048,
+                                return_tensors='pt')
+        return encoding
+
+    def generate_program(self, seed: str) -> dict:
+        model, tokenizer = model_setup(self.cfg)
+        encoding = self.generate_prompt_str(seed, tokenizer)
+        completion = sample(self.cfg, model, tokenizer, encoding)
+        truncation = truncate(completion)
```
Comment (honglu2875) [inline diff_model.py:84]: Got an error here. Looks like `completion` is a list of str and truncation works on a single str.
`truncation = list(map(truncate, completion))` (and some changes to the next few lines) is working for me.

**Comment by daia99** [comment]
https://github.com/CarperAI/ELM/blob/6bf347c13301afadfb093616a6b84d64a59a686f/map_elites/map_elites.py#L59-L62

To support replicating the datasets for Stage 2, we'll need to record all examples mapped to the archive.

Not sure if this is within scope of the stage 1 commit, but what's a good way to add support for this option (for ELM), while keeping this more general as a map_elites implementation? Would this make sense as a parameter/option in the env that can be checked?

**Comment by herbiebradley** [comment]
Merged to get basic functionality into main. Will make a PR to add code style requirements, documentation, etc soon.

**Pull request merged**